File: 数组/螺旋矩阵II.go
package 数组

//用1～n^2以螺旋的形式填满矩阵
func generateMatrix(n int) [][]int {
	mat := make([][]int, n)
	for i := 0; i < n; i++ {
		mat[i] = make([]int, n)
	}
	var nums []int
	for i := 2; i <= n*n; i++ {
		nums = append(nums, i)
	}
	var col int
	var row int
	var nextSetp func(int, int)

	nextSetp = func(i, j int) {
		if j+1 < n && mat[i][j+1] == 0 { ////right
			for col = j + 1; col < n && mat[i][col] == 0; col++ {
				mat[i][col] = nums[0]
				nums = append(nums[:0], nums[1:]...)
			}
			col--
			nextSetp(i, col)
		}
		if j-1 >= 0 && mat[i][j-1] == 0 { //left
			for col = j - 1; col >= 0 && mat[i][col] == 0; col-- {
				mat[i][col] = nums[0]
				nums = append(nums[:0], nums[1:]...)
			}
			col++
			nextSetp(i, col)
		}
		if i+1 < n && mat[i+1][j] == 0 { //down
			for row = i + 1; row < n && mat[row][j] == 0; row++ {
				mat[row][j] = nums[0]
				nums = append(nums[:0], nums[1:]...)
			}
			row--
			nextSetp(row, j)
		}
		if i-1 >= 0 && mat[i-1][j] == 0 { //up
			for row = i - 1; row >= 0 && mat[row][j] == 0; row-- {
				mat[row][j] = nums[0]
				nums = append(nums[:0], nums[1:]...)
			}
			row++
			nextSetp(row, j)
		}
	}
	mat[0][0] = 1
	nextSetp(0, 0)
	return mat
}

File: 模拟/螺旋矩阵II.java
package 模拟;

//以递归的方式进行模拟

public class 螺旋矩阵II {
    static int[][] mat;

    public static int[][] generateMatrix(int n) {
        mat = new int[n][n];
        mat[0][0] = 1;
        //我目前在（0，0）位置，我下一个要填入的数字是2，接下来怎么走交给你了
        nextStep(0, 0, 2);
        return mat;
    }

    public static void nextStep(int row, int col, int nextNum) {
        int n = mat.length;
        if (col + 1 < n && mat[row][col + 1] == 0) {
            int j = col + 1;
            for (j = col + 1; j < n && mat[row][j] == 0; j++) {
                mat[row][j] = nextNum++;
            }

            nextStep(row, j - 1, nextNum);
        }
        if (row + 1 < n && mat[row + 1][col] == 0) {
            int i = row + 1;
            for (; i < n && mat[i][col] == 0; i++) {
                mat[i][col] = nextNum++;
            }
            nextStep(i - 1, col, nextNum);
        }
        if (col - 1 >= 0 && mat[row][col - 1] == 0) {
            int j = col - 1;
            for (; j >= 0 && mat[row][j] == 0; j--) {
                mat[row][j] = nextNum++;
            }
            nextStep(row, j + 1, nextNum);
        }
        if (row - 1 >= 0 && mat[row - 1][col] == 0) {
            int i = row - 1;
            for (; i >= 0 && mat[i][col] == 0; i--) {
                mat[i][col] = nextNum++;
            }
            nextStep(i + 1, col, nextNum);
        }
    }

    public static void main(String[] args) {
        generateMatrix(3);
    }
}
